integration/restValidation.ts

```diff
@@ -0,0 +1,162 @@
+        [...validators].some(
+            (index) => index === validatorIndex && (slot > ignoreBefore || ignoreBefore === undefined),
+        ),
+    );
+};
+
+const getProposerFactory = (apiClient: ICgEth2ApiClient) => async (
+    pubKey: BLSPubkey,
+    epoch: number,
+): Promise<ProposerDuty[]> => await apiClient.validator.getProposerDuties(epoch, [pubKey]);
+
+const isBlockProposedFactory = (apiClient: ICgEth2ApiClient) => async (slot: number): Promise<boolean> =>
+    !!(await apiClient.beacon.blocks.getBlockAttestations(slot));
+
+export const restValidation = ({
+    baseUrl,
+    getValidatorPrivateKey,
+    limit,
+}: {
+    baseUrl: string;
+    getValidatorPrivateKey: () => Promise<SecretKey>;
+    limit: number;
+}): Promise<{
+    proposer: {
+        proposed: number;
+        delegated: number;
+    };
+    attestation: {
+        attestations: number;
+        delegated: number;
+    };
+}> =>
+    new Promise((resolve) => {
+        (async (): Promise<void> => {
+            await initBLS("blst-native");
+            const validatorPrivateKey = await getValidatorPrivateKey();
+            const validatorPublicKey = validatorPrivateKey.toPublicKey();
+            const validatorPublicKeyBytes = validatorPublicKey.toBytes();
+            console.log("Starting validator " + validatorPublicKey.toHex());
+
+            const logger = new WinstonLogger({module: "ChainGuardian", level: LogLevel.verbose});
+            const slashingProtection = sinon.createStubInstance(CGSlashingProtection);
+
+            const eth2API = new CgEth2ApiClient(config, baseUrl);
+            const validatorService = new Validator({
+                slashingProtection,
+                api: eth2API,
+                config,
+                secretKeys: [validatorPrivateKey],
+                logger,
+                graffiti: "ChainGuardian",
+            });
+            const validator = await eth2API.beacon.state.getStateValidator("head", validatorPublicKeyBytes);
+            await validatorService.start();
+
+            let firstSlot: number | undefined;
+            let slot: number | undefined;
+
+            let startEpoch = 1;
+            let lastEpoch = 1;
+
+            const getCommittees = getCommitteesFactory(eth2API);
+            let committees: BeaconCommitteeResponse[] = [];
+            const attestations = new Map<number, boolean>();
+
+            const getProposer = getProposerFactory(eth2API);
+            const isBlockProposed = isBlockProposedFactory(eth2API);
+            const proposers = new Map<number, boolean>();
+
+            const stream = await eth2API.events.getEventStream(([
+                CGBeaconEventType.HEAD,
+                CGBeaconEventType.ATTESTATION,
+            ] as unknown) as BeaconEventType[]);
+            for await (const {type, message} of stream) {
+                if (((type as unknown) as CGBeaconEventType.ATTESTATION) === CGBeaconEventType.ATTESTATION) {
```

Comment:
to make it more readable, could you move each case implementation to separate function?

Reply:
it's hard to do that, as this cases depends on the upper scope

Reply:
Can't you pass required params as method args?

Reply:
`CGBeaconEventType.ATTESTATION` just use Map `attestations`
but `CGBeaconEventType.HEAD` uses from uper scope
- slot
- proposers
- firstSlot
- startEpoch
- lastEpoch
- committees
- proposers
- attestations  

pretty needy case :see_no_evil:

Reply:
pass everything as huge object. It's really hard to follow that huge piece of code 😅

Reply:
and then return as the very huge object to assign all variables to upper scope variables?